### ISSUE TITLE
cannot override None in list

### ISSUE THREAD

**Issue by nschloe** [microsoft/pyright#1229]
MWE:
```python
p = [None]
p[0] = 1  # !
```
```
  3:1 - error: No overloads for "p[0]" match parameters
    Calling function "__setitem__"
    Argument types: (Literal[0], Literal[1]) (reportGeneralTypeIssues)
```
Same for
```python
p = [1]
p[0] = None
```

**Comment by erictraut** [comment]
This is the correct (intended) behavior. Because you haven't provided a type declaration for `p`, the type checker infers the type based on the assignment. For details about pyright's inference rules for list expressions, refer to [this documentation](https://github.com/microsoft/pyright/blob/master/docs/type-inference.md#list-expressions).

The workaround in this case is to provide a type annotation that declares the type of `p` as follows:

```python
p: List[Optional[int]] = [None]
```

**Comment by nschloe** [comment]
Thanks for the explanation!
